### PR TITLE
[MI-2505][server]: Added logic to protect subscriptions notification webhook API and fixed Boards update subscription.

### PR DESCRIPTION
**Pull request by avas27JTG**
- Added logic to protect subscriptions notification webhook API
- Fixed Boards update subscription.

### CODE REVIEW

File: plugin.json [hunk 1]
```diff
@@ -44,6 +44,12 @@
                 "display_name": "Encryption Secret:",
                 "type": "generated",
                 "help_text": "The secret key used to encrypt and decrypt the OAuth token.\nRegenerating the secret will require all users to re-connect their accounts to Azure DevOps."
+            },
+            {
+                "key": "WebhookSecret",
+                "display_name": "Webhook Secret:",
+                "type": "generated",
+                "help_text": "The secret key used to authenticate incoming requests to notification webhook used for subscriptions\nRegenerating the secret will require all users to create new subscriptions as existing ones will no longer be authenticated."
```
Comment (ayusht2810) [inline plugin.json:52]:
```suggestion
                "help_text": "The secret key used to authenticate incoming requests to notification webhook used for subscriptions.\nRegenerating the secret will require all users to create new subscriptions as existing ones will no longer be authenticated."
```

File: server/config/config.go [hunk 2]
```diff
@@ -57,6 +59,9 @@ func (c *Configuration) IsValid() error {
 	if c.EncryptionSecret == "" {
 		return errors.New(constants.EmptyEncryptionSecretError)
 	}
+	if c.WebhookSecret == "" {
+		return errors.New(constants.WebhookSecretError)
```
Comment (ayusht2810) [inline server/config/config.go:63]: we can name this variable similar to other ones `EmptyWebhookSecretError`

File: server/config/config_test.go [hunk 3]
```diff
@@ -22,6 +22,7 @@ func TestIsValid(t *testing.T) {
 				AzureDevopsOAuthAppID:        "mockAzureDevopsOAuthAppID",
 				AzureDevopsOAuthClientSecret: "mockAzureDevopsOAuthClientSecret",
 				EncryptionSecret:             "mockEncryptionSecret",
+				WebhookSecret:                "mockWebhookSecret",
 			},
 			errMsg: "",
```
Comment (ayusht2810) [inline server/config/config_test.go:27]:
```suggestion
```

File: server/constants/constants.go [hunk 4]
```diff
@@ -95,6 +95,10 @@ const (
 	GetTasks    = "/%s/_apis/wit/workitems"
 	StaticFiles = "%s/plugins/%s/static/%s"
 
+	// Query params sent to Azure DevOps APIs
```
Comment (ayusht2810) [inline server/constants/constants.go:98]:
```suggestion
	// Query params sent from Azure DevOps APIs
```

File: server/plugin/api.go [hunk 5]
```diff
@@ -489,14 +489,21 @@ func (p *Plugin) getReviewersListString(reviewersList []serializers.Reviewer) st
 }
 
 func (p *Plugin) handleSubscriptionNotifications(w http.ResponseWriter, r *http.Request) {
+	webhookSecret := r.URL.Query().Get(constants.AzureDevopsQueryParamWebhookSecret)
+	if webhookSecret == "" || webhookSecret != p.getConfiguration().WebhookSecret {
```
Comment (ayusht2810) [inline server/plugin/api.go:493]: can we use this here https://github.com/Brightscout/mattermost-plugin-servicenow/blob/e9d9ae34f20bc52872e7ab022924b582208b4ea1/server/plugin/api.go#L824

File: server/constants/constants.go (hunk 4 above)
Comment (ayusht2810) [reply server/constants/constants.go:98]: shouldn't it be this as we are receiving requests?

File: server/plugin/api_test.go [hunk 6]
```diff
@@ -763,13 +769,15 @@ func TestHandleSubscriptionNotifications(t *testing.T) {
...
 		},
 		{
-			description: "SubscriptionNotifications: invalid channel ID",
-			body:        `{}`,
-			err:         errors.New("mockError"),
-			channelID:   "mockChannelID",
-			statusCode:  http.StatusBadRequest,
+			description:   "SubscriptionNotifications: invalid channel ID",
+			body:          `{}`,
+			err:           errors.New("mockError"),
```
Comment (ayusht2810) [inline server/plugin/api_test.go:777]: write some meaningful error here

File: server/plugin/api_test.go [hunk 7]
```diff
@@ -800,7 +820,7 @@ func TestHandleSubscriptionNotifications(t *testing.T) {
 				return testCase.isValidChannelID
 			})
 
-			req := httptest.NewRequest(http.MethodPost, fmt.Sprintf("/notification?channelID=%s", testCase.channelID), bytes.NewBufferString(testCase.body))
+			req := httptest.NewRequest(http.MethodPost, fmt.Sprintf("/notification?channelID=%s&webhookSecret=%s", testCase.channelID, testCase.webhookSecret), bytes.NewBufferString(testCase.body))
```
Comment (ayusht2810) [inline server/plugin/api_test.go:823]: use constant for the path if present

File: server/constants/messages.go [hunk 8]
```diff
@@ -24,53 +24,56 @@ const (
...
 	ServiceTypeRequired             = "service type is required"
 	ChannelIDRequired               = "channel ID is required"
+	WebhookSecretRequired           = "webhook secret is required"
 	MMUserIDRequired                = "mattermsot user ID is required"
 	EmptyAzureDevopsAPIBaseURLError = "azure devops API base URL should not be empty"
 	EmptyAzureDevopsOAuthAppIDError = "azure devops OAuth app id should not be empty"
 	// #nosec G101 -- This is a false positive. The below line is not a hardcoded credential
 	EmptyAzureDevopsOAuthClientSecretError = "azure devops OAuth client secret should not be empty"
 	EmptyEncryptionSecretError             = "encryption secret should not be empty"
+	WebhookSecretError                     = "webhook secret should not be empty"
```
Comment (raghavaggarwal2308) [inline server/constants/messages.go:34]:
```suggestion
	EmptyWebhookSecretError                     = "webhook secret should not be empty"
```
Just for consistancy

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/constants/constants.go (hunk 4 above)
Comment (avas27JTG) [reply server/constants/constants.go:98]: No, we are sending these to Azure

File: server/plugin/api.go (hunk 5 above)
Comment (avas27JTG) [reply server/plugin/api.go:493]: have to change some logic as here we are receiving as query param and not URL encoded FORM

File: server/config/config_test.go (hunk 3 above)
Comment (avas27JTG) [reply server/config/config_test.go:27]: giving error on removing it @raghavaggarwal2308 can look into it

Comment (raghavaggarwal2308) [reply server/config/config_test.go:27]: @avas27JTG Fixed [here](https://github.com/Brightscout/mattermost-plugin-azure-devops/pull/22#pullrequestreview-1243581764)

**Pull request closed without merging**